### PR TITLE
mushr-unity-sim tutorial

### PR DESCRIPTION
**Pull request by naughtyStark**
I've put in the content necessary to get started with using the mushr_unity_sim backend. links to the repository will be updated when the mushr_unity_sim package gets a go-ahead for transfer to prl-mushr.

### CODE REVIEW

File: content/tutorials/mushr_unity_sim.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+---
+title: "Using unity to simulate MuSHR"
+date: 2021-05-17T17:51:18+05:30
+summary: "Using a unity-based simulator in place of the default state prediction system that comes with mushr_base"
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:4]: Using a unity-based physics simulator in place of the default kinematic mushr_sim

File: content/tutorials/mushr_unity_sim.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+active: true     # whether this is listed at /tutorials/
+draft: true      # whether Hugo considers this a draft
+weight: 3        # 2 = intro tutorial 3 = anything else
+---
+
+<h2> By: <a href=https://www.sidharthtalia.com/>Sidharth Talia</a></h2>
+
+<!-- Header figure required! -->
+<br>
+{{< figure src="/tutorials/mushr_unity_sim/thumbnail.gif" width="800" >}} <br>
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:17]: good gif, can you zoom and crop so it is only the rviz window?

File: content/tutorials/mushr_unity_sim.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+<!-- Header figure required! -->
+<br>
+{{< figure src="/tutorials/mushr_unity_sim/thumbnail.gif" width="800" >}} <br>                           
+<br>
+
+### Introduction
+The default state prediction system [racecar_state.py](https://github.com/prl-mushr/mushr_base/blob/master/mushr_base/src/racecar_state.py) works well for low-speed situations where the wheel-slippage and body roll aren't significant. However, once you get to higher speeds (3-4 m/s and above), you want to be able to simulate those things as well so that you know your system is robust to slippage, body roll and so on. For this purpose, we use a unity-based simulator to simulate the state of the car with higher physics fidelity. The system provided by the [mushr_unity_sim](https://github.com/naughtyStark/mushr_unity_sim.git) is a drop in replacement for the default state predictor.
+
+<br>
+{{< figure src="/tutorials/mushr_unity_sim/thumbnail.png" width="800" >}} <br>
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:24]: oh super cool pic!

File: content/tutorials/mushr_unity_sim.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+<h2> By: <a href=https://www.sidharthtalia.com/>Sidharth Talia</a></h2>
+
+<!-- Header figure required! -->
+<br>
+{{< figure src="/tutorials/mushr_unity_sim/thumbnail.gif" width="800" >}} <br>                           
+<br>
+
+### Introduction
+The default state prediction system [racecar_state.py](https://github.com/prl-mushr/mushr_base/blob/master/mushr_base/src/racecar_state.py) works well for low-speed situations where the wheel-slippage and body roll aren't significant. However, once you get to higher speeds (3-4 m/s and above), you want to be able to simulate those things as well so that you know your system is robust to slippage, body roll and so on. For this purpose, we use a unity-based simulator to simulate the state of the car with higher physics fidelity. The system provided by the [mushr_unity_sim](https://github.com/naughtyStark/mushr_unity_sim.git) is a drop in replacement for the default state predictor.
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:21]: default state predictionn system racecar_state.py used in mushr_sim

Comment (schmittlema) [reply content/tutorials/mushr_unity_sim.md:21]: mention unity's underlying physics engine. I think it is bullet?

File: content/tutorials/mushr_unity_sim.md (hunk 3 above)
Comment (schmittlema) [reply content/tutorials/mushr_unity_sim.md:24]: maybe this is better suited for the nhttc tutorial though?

File: content/tutorials/mushr_unity_sim.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+$ cd ~/catkin_ws
+$ catkin_make
+{{</highlight>}}
+
+The last catkin_make command is necessary to make the system recognize mushr_unity_sim as a valid ros package.
+
+Python dependencies: 
+
+{{<highlight bash>}}
+$ pip install -r requirements.txt
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:54]: `cd src/mushr_unity_sim`

File: content/tutorials/mushr_unity_sim.md (hunk 4 above)
Comment (schmittlema) [reply content/tutorials/mushr_unity_sim.md:21]: It seems you should give credit to donkey sim no?

File: content/tutorials/mushr_unity_sim.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+$ ./donkey_sim.x86_64 -batchmode
+{{</highlight>}}
+
+If you have nvidia driver support on your linux machine (god bless), you can run it without the "-batchmode" tag. The tag makes the sim run in the headless mode which allows for higher fps if you don't have said driver support. Since the simulation surroundings/environment have been simplified to reduce the size of the simulator, there isn't much to see except the car itself and other cars if they are spawned.
+
+
+In a new tab:
+
+{{<highlight bash>}}
+$ roslaunch mushr_unity_sim unity_multi.launch
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:72]: how do I launch 1 car?

File: content/tutorials/mushr_unity_sim.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+{{</highlight>}}
+In another new tab:
+{{<highlight bash>}}
+$ rviz -d ~/catkin_ws/src/mushr_unity_sim/rviz/mushr_unity_sim.rviz
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:76]: rviz config is empty

File: content/tutorials/mushr_unity_sim.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+{{</highlight>}}
+
+You should see 4 cars by default. The poses of these cars are set by the pose_init.py file. The car's initial pose is set the same way as done for the default mushr_sim; by publishing a message on the topic: `/car_name/initialpose` of type: [geometry_msgs/PoseWithCovarianceStamped](http://docs.ros.org/en/melodic/api/geometry_msgs/html/msg/PoseWithCovarianceStamped.html) where `car_name` corresponds to the name of the car. 
+
+
+### Using this simulator for your own purposes:
+In order to use this simulation backend in your system, all you need to do is replace the default call to the single_car.launch (as shown below)
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:83]: default call in mushr_sim? if so link the file

File: content/tutorials/mushr_unity_sim.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+{{<highlight xml>}}
+    <group ns="$(arg car_name)">
+        <include file="$(find mushr_unity_sim)/launch/single_car.launch">
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:86]: mushr_unity_sim --> mushr_sim ???

File: content/tutorials/mushr_unity_sim.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+$ pip install -r requirements.txt
+{{</highlight>}}
+
+
+### Running the base example:
+In a new terminal, enter:
+
+{{<highlight bash>}}
+$ cd ~/catkin_ws/src/mushr_unity_sim/unity
+$ ./donkey_sim.x86_64 -batchmode
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:63]: could you run this through the launchfile?

File: content/tutorials/mushr_unity_sim.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+{{<highlight bash>}}
+$ roslaunch mushr_unity_sim unity_multi.launch
+{{</highlight>}}
+In another new tab:
+{{<highlight bash>}}
+$ rviz -d ~/catkin_ws/src/mushr_unity_sim/rviz/mushr_unity_sim.rviz
+{{</highlight>}}
+
+You should see 4 cars by default. The poses of these cars are set by the pose_init.py file. The car's initial pose is set the same way as done for the default mushr_sim; by publishing a message on the topic: `/car_name/initialpose` of type: [geometry_msgs/PoseWithCovarianceStamped](http://docs.ros.org/en/melodic/api/geometry_msgs/html/msg/PoseWithCovarianceStamped.html) where `car_name` corresponds to the name of the car.
```
Comment (schmittlema) [inline content/tutorials/mushr_unity_sim.md:79]: mention driving around and what they should see differently

Comment (schmittlema) [reply content/tutorials/mushr_unity_sim.md:79]: the car slips a lot is that to be expected?

**Review by schmittlema — CHANGES_REQUESTED** [review]
looking pretty good! various small things throughout

File: content/tutorials/mushr_unity_sim.md (hunk 1 above)
Comment (naughtyStark) [reply content/tutorials/mushr_unity_sim.md:4]: yeah, I did feel it was a bit wordy.

File: content/tutorials/mushr_unity_sim.md (hunk 2 above)
Comment (naughtyStark) [reply content/tutorials/mushr_unity_sim.md:17]: Will do.

File: content/tutorials/mushr_unity_sim.md (hunk 3 above)
Comment (naughtyStark) [reply content/tutorials/mushr_unity_sim.md:24]: you're probably right. I wanted to have a "mean looking" line-up somewhere. I can certainly put this in the nhttc tutorial. I'll put a different gif here.

File: content/tutorials/mushr_unity_sim.md (hunk 4 above)
Comment (naughtyStark) [reply content/tutorials/mushr_unity_sim.md:21]: Yes. Forgot to put links to all those things. I'll confirm what physics engine is being used.

File: content/tutorials/mushr_unity_sim.md (hunk 10 above)
Comment (naughtyStark) [reply content/tutorials/mushr_unity_sim.md:63]: I looked into this a while ago and I gave up on it for some reason. I'll look into this again.

File: content/tutorials/mushr_unity_sim.md (hunk 11 above)
Comment (naughtyStark) [reply content/tutorials/mushr_unity_sim.md:79]: actually, it's good that you brought that up. I modeled the car's behavior based on how it drove on the surface of my roof, which is smoother than tarmac. I'm not sure which surface you guys drive on and I've not driven on that surface. Could someone from the lab show me what the car looks like when its turning at the limit of it's grip?

File: content/tutorials/mushr_unity_sim.md (hunk 4 above)
Comment (naughtyStark) [reply content/tutorials/mushr_unity_sim.md:21]: I believe they use PhysX from Nvidia.

File: content/tutorials/mushr_unity_sim.md (hunk 10 above)
Comment (naughtyStark) [reply content/tutorials/mushr_unity_sim.md]: This now runs from the launch file
